File: course1/week4/2_majority_element/python/main.py
# python3

##
#
# Python3 implementation of majority element
#
# (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
#
##

from typing import List

class Solution:
    def hasMajorityElement( self, A: List[int], N: int ) -> bool:
        me = 0
        cnt = 0
        for x in A:
            if cnt == 0:
                me = x
            cnt = cnt + 1 if me == x else cnt - 1
        cnt = 0
        T = ( N // 2 ) + 1
        for x in A:
            cnt = cnt + 1 if me == x else cnt
        return T <= cnt

if __name__ == '__main__':
    solution = Solution()
    N = int( input() )
    A = list( map( int, input().split() ))
    ans = solution.hasMajorityElement( A, N )
    print( "1" if ans else "0" )

File: course1/week4/2_majority_element/java/src/Main.java
/**
 *
 * Java implementation of majority element
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

import java.util.Scanner;

public class Main {

    public static boolean hasMajorityElement( int[] A, int N ){
        int me = 0;
        for( int i=( 0 ), cnt=( 0 ); i < N; cnt=(( me == A[ i++ ] ) ? ( cnt + 1 ) : ( cnt - 1 )) )
            if( cnt == 0 )
                me = A[ i ];
        int cnt = 0,
            T = ( N / 2 ) + 1;
        for( int i=( 0 ); i < N; ++i )
            if( me == A[ i ] )
                ++cnt;
        return( T <= cnt );
    }

    public static void main( String[] args ){
        Scanner input = new Scanner( System.in );
        int N = input.nextInt();
        int[] A = new int[ N ];
        for( int i=( 0 ); i < N; A[ i++ ]=( input.nextInt() ));
        boolean ans = hasMajorityElement( A, N );
        System.out.println( ans? "1" : "0" );
    }
}
